lib/Sema/CSDiag.cpp

```diff
@@ -5730,10 +5730,32 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
       return true;
 
     if (!lhsType->isEqual(rhsType)) {
-      diagnose(callExpr->getLoc(), diag::cannot_apply_binop_to_args,
-               overloadName, lhsType, rhsType)
-      .highlight(lhsExpr->getSourceRange())
+      auto diag = diagnose(callExpr->getLoc(), diag::cannot_apply_binop_to_args,
+                           overloadName, lhsType, rhsType);
+      diag.highlight(lhsExpr->getSourceRange())
       .highlight(rhsExpr->getSourceRange());
+
+      for (auto &candidate : calleeInfo.candidates) {
+        auto tupleType = dyn_cast<TupleType>(candidate.getArgumentType().getPointer());
```

Comment:
It would be better to use:
```c++
if (auto *fnType = candidate.getUncurriedFunctionType()) {
   auto params = fnType->getParams();
}
```
here since we are trying to move away from using parameter types directly.

Reply:
The problem though is that even though we _could_ diagnose one of the candidates with cast fix-it it doesn't necessarily mean that we _should_ do that... E.g. if I had `+(A, Int) -> Float` and `+(A, Double) -> Float` and I have `A() + x` where `x: Float` expression like in my previous example, the question is which fix-it should I suggest, `Int` or `Double`? IMHO none, because it's ambiguous what the user wanted to do here...

Reply:
So what is supposed to do in following case?

```
let x: Int = 0
let y: Float = 0
x * y
```

We have 2 candidates to suggest "fix-it".

```
x * Int(y)
```

or 

```
Float(x) * y
```

You think we should not suggest in this case?

Reply:
Yes, that's right, I don't think it makes sense to suggest fix-it in this case because it's unclear what is intended.

Reply:
Addressed!
https://github.com/apple/swift/pull/13272/commits/d80501f7346919b0b7e440d4ad4aa739b49fbca8

Reply:
Addressed https://github.com/apple/swift/pull/13272/commits/0df987a4ff537cce94e8a0ebad917d739e94e3d2